packages/dd-trace/src/datastreams/processor.js

```diff
@@ -49,6 +53,26 @@ class StatsBucket extends Map {
   }
 }
 
+function calculateByteSize (data) {
```

Comment:
I was curious what type information was exposed by `kafkajs`, so I did a little digging. Since [v1.10.0](https://github.com/tulios/kafkajs/blob/master/types/index.d.ts#L109-L115), this is:

```ts
export interface Message {
  key?: Buffer | string | null
  value: Buffer | string | null
  partition?: number
  headers?: IHeaders
  timestamp?: string
}
```

Previously, types were provided by DefinitelyTyped ([producer](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f08baa1d0ffd812aabc478aedeb8dfa6fdbc2059/types/kafkajs/index.d.ts#L193-L198), [consumer](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f08baa1d0ffd812aabc478aedeb8dfa6fdbc2059/types/kafkajs/index.d.ts#L306-L312)):

```ts
export interface ProducerMessage {
    partition?: number;
    key?: string;
    value: string | Buffer | ArrayBuffer;
    headers?: { [key: string]: string };
}
```

```ts
export interface ConsumerMessage {
    timestamp: number;
    key: string;
    value: Buffer;
    headers: { [key: string]: string };
    offset: number;
}
```

The only mildly annoying option here is that producer message values may be `ArrayBuffer`s. 
___

Following the [.net implementation](https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Kafka/KafkaHelper.cs#L100-L128) and other tidbits, we can do a few things:
* JS strings are UTF-16 (or UCS-2), so we need `Buffer.from(str).length` (default encoding is UTF-8 here, but we can be explicit to show intent)
* Conversely, we need to assume something for `Buffer` args, UTF-8 is sane so just `buf.length` - spot this using [`Buffer.isBuffer()`](https://nodejs.org/api/buffer.html#static-method-bufferisbufferobj)
* There are no `Array`s, only `ArrayBuffer`s, which we can spot using [isArrayBuffer](https://nodejs.org/api/util.html#utiltypesisarraybuffervalue). Best-effort, IMHO, is `arrayBuffer.byteLength`.
* We forgot to return 0 if we don't support a type
* I like dotnet's separation between leaf types and structural information so I'm replicating it, but YMMV

```js
function getSizeOrZero(obj) {
  if (typeof obj === 'string') {
    return Buffer.from(obj, 'utf-8').length
  }
  if (types.isArrayBuffer(obj) {
    return obj.byteLength
  }
  if (Buffer.isBuffer(obj) {
    return obj.length
  }
  return 0
}

function getHeadersSize(headers) {
  if (headers === undefined) return 0
  return Object.entries(headers).reduce(([key, val], prev) => getSizeOrZero(key) + getSizeOrZero(val) + prev, 0)
}

function getMessageSize(message) {
  const { key, value, headers } = message
  return getSizeOrZero(key) + getSizeOrZero(value) + getHeadersSize(headers)
}

function getPayloadSize(payload) {
  return payload.reduce(getMessageSize, 0)
}
```

Reply:
Updated the PR to use this functionality